Update LHN/chat header to handle displayName not being set

lukemorawski opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/27393 
PROPOSAL: No proposal


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Open the app (prefferably in a web browser) 
- [ ] Open dev tools and take a look in to the IndexedDB (Application tab in Chrome Dev Tools)
- [ ] In `personalDetailsList` key, find the object that represents personal details of a contact you want to make hidden
- [ ] In the dev console punch in the following command:
```
Onyx.merge('personalDetailsList', {<account id you want to change>: {accountID: <account id you want to change>, avatar: "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/default-avatar_24.png", displayName: "", firstName: "", lastName: "", login: "", phoneNumber: "", validated: true}})
```
- [ ] Observe the LHN and look for "Hidden" instead of the name of the contact
- [ ] Open the chat with that contact
- [ ] Put the above command in the console again (Onyx probably overwrote the data with real ones by now)
- [ ] Observe the header changing name to Hidden
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
<img src="https://github.com/Expensify/App/assets/134388952/4dd286a6-7e99-4686-9674-faa03ed5c139" />


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<img src="https://github.com/Expensify/App/assets/134388952/dac64a52-bb03-4217-9c4f-ac47f212d971" />

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>
<img src="https://github.com/Expensify/App/assets/134388952/9c7f16b5-6ee9-4577-9980-eb51e8e3ed9a" />
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>
<img src="https://github.com/Expensify/App/assets/134388952/3b920602-69c0-47f3-80b6-aa4993eb43f0" />

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>
<img src="https://github.com/Expensify/App/assets/134388952/fe3c5936-b0dd-48a8-bedf-4a50f42ca5eb" />

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>
<img src="https://github.com/Expensify/App/assets/134388952/95452bad-a9ca-4ace-83ec-9adac4f37c1f" />

<!-- add screenshots or videos here -->

</details>

lukemorawski committed:
displayName fallback to Hidden

puneetlath commented:
@mananjadhav hi! Do you have an ETA for reviewing?

mananjadhav commented:
Will be reviewing this today within 2-3 hours.

mananjadhav commented:
@puneetlath @lukemorawski The `Hidden` works as expected but I am not sure if we should be showing hidden when we search for a new user.

Steps to reproduce:
1. Open the App
2. Click on the `Search` icon
3. Enter the email of the user you've never chatted before.
4. Check the name of the user, and you'll see it shows `Hidden`

<img width="1512" alt="web-hidden-in-search" src="https://github.com/Expensify/App/assets/3069065/96ab88e4-9532-4974-b850-f3c27ef0454e">

puneetlath commented:
Ah interesting. Yeah on the search page, let's just show the email address as the display name like we do today. Would you be able to make that update @lukemorawski?

puneetlath commented:
@lukemorawski think you'll be able to make that update today?

lukemorawski commented:
@puneetlath yep!

lukemorawski committed:
fix/search list shows hidden for unkown yet contact

lukemorawski commented:
@puneetlath Added some changes. Please check :)

puneetlath commented:
Thanks! @mananjadhav mind giving it a quick look first?

mananjadhav commented:
@lukemorawski Now for the valid ones too, we're seeing a blank. 

1. Create an iOU request with a user
2. Set it's value as empty based on the `Onyx.merge` command
3. Open the search bar and look for the user. You'll see the title shows blank. 


<img width="423" alt="Screenshot 2023-09-26 at 12 57 52 AM" src="https://github.com/Expensify/App/assets/3069065/064cf766-c8ab-494c-8548-6641e5ffa444">


https://github.com/Expensify/App/assets/3069065/f6a96b89-6349-4bce-bb5d-6a52836e136a

@puneetlath I think we should be able to differentiate between the new chat or existing hidden records. Can't backend send some flag from the backend, where we want to show Hidden?

puneetlath commented:
>Can't backend send some flag from the backend, where we want to show Hidden?

It's not really so simple. The back-end will send the data that the user should have access to. Which sometimes will be just accountID. Sometimes will be accountID and displayName. Sometimes will be accountID, displayName, and login. 

Can we not do this based on what information we have locally?

puneetlath commented:
@lukemorawski what are your thoughts? Would love to get this wrapped up this week!

lukemorawski commented:
@puneetlath Sorry, had some more urgent tickets. Will check this out today!

lukemorawski commented:
@puneetlath @mananjadhav tricky thing, but there might be a way to deal with that locally

lukemorawski committed:
more fixes

lukemorawski commented:
@puneetlath @mananjadhav OK, made some fixes. Looks good for me, but please do check it out!

puneetlath commented:
@mananjadhav thoughts on the approach?

mananjadhav commented on src/libs/ReportUtils.js:
```diff
@@ -1089,32 +1089,46 @@ function getPersonalDetailsForAccountID(accountID) {
  *
  * @param {Number} accountID
  * @param {Boolean} [shouldUseShortForm]
+ * @param {Boolean} shouldNotFallbackToHidden
```
Agreed with @rushatgabhane here.

lukemorawski replied:
ok, though it made more sense like that, as falling back to hidden is the default :)

mananjadhav commented:
@lukemorawski quick bump on the changes requested.

lukemorawski commented on src/libs/ReportUtils.js:
```diff
@@ -1089,32 +1089,46 @@ function getPersonalDetailsForAccountID(accountID) {
     const personalDetails = getPersonalDetailsForAccountID(accountID);
+    // check if it's invite account
+    if (hasOnlyAvatarField(personalDetails)) {
+        return '';
```
Sure! Say a user is entering an email address into the search field, and that address has not yet been registered as Expensify user, it will, in that scenario have no other details than automatically created avatar. This is how I detect a "new invite account", because the can't fallback to "Hidden" `displayName` value.

mananjadhav replied:
Thanks for the explanation @lukemorawski. Let's add a comment as a JSDoc for the method.

lukemorawski commented on src/libs/ReportUtils.js:
```diff
@@ -1089,32 +1089,46 @@ function getPersonalDetailsForAccountID(accountID) {
 
+function hasOnlyAvatarField(obj) {
+    const keys = _.keys(obj);
+    return keys.length === 1 && keys[0] === 'avatar';
```
not sure why. This is safe and very readable. Lodash would be something like this
```
_.isEqual(_.keys(obj), ['avatar']);
```

puneetlath replied:
What happens to this personalDetails object post-invite once the API responds. Does it get cleared out? If so, perhaps we can add an explicit `isOptimistc` key to it instead of having to infer it.

mananjadhav replied:
Not sure what you mean by `cleared out`? But when we search for a non-existing account we see this.
<img width="878" alt="image" src="https://github.com/Expensify/App/assets/3069065/06e3f13b-2572-4dde-af04-db6c25d42cac">

Once we receive the object again post-invite, it includes all the keys. For example,

```
{
    "accountID": 15820244,
    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/default-avatar_4.png",
    "displayName": "fwefwewfwe@wefnkewn.com",
    "firstName": "",
    "lastName": "",
    "status": null
}

```

puneetlath replied:
Sorry, I think I'm confused by what an "invite" account is. 

From what I can see, when I invite a new user, we create an optimistic personalDetails object for that user that looks like this: 
<img width="633" alt="Screenshot 2023-10-13 at 10 35 27 AM" src="https://github.com/Expensify/App/assets/1371996/33c23b9a-0f13-470d-b451-2eddad61517e">

Then when the API responds, the optimistic personalDetail gets removed and the real personalDetail with the real accountID for that user is added: 
<img width="594" alt="image" src="https://github.com/Expensify/App/assets/1371996/c7a95278-2d51-4d43-9cba-ef79cd41a071">

So I'm just trying to understand what this "invite" account that has only the avatar field is and when it would exist. Sorry if I'm missing something obvious!

lukemorawski commented on src/libs/ReportUtils.js:
```diff
@@ -1089,32 +1089,46 @@ function getPersonalDetailsForAccountID(accountID) {
     );
 }
 
+function hasOnlyAvatarField(obj) {
```
answered above

mananjadhav replied:
Let's add a comment as a JSDoc for the method.

lukemorawski committed:
small refactors

lukemorawski commented:
@mananjadhav made some refactoring per your requests :)

lukemorawski committed:
Merge branch 'main' into 27393-displayName_fallback_to_hidden

mananjadhav commented:
Did further testing. I have few bugs/questions overall. @puneetlath Would be great if you can also point out the ones we might not require. Also do we have a real account that we can replicate for this behavior? Basically updating the Onyx every now and then doesn't seem very reliable. Difficult to identify the cases it might throw a console error too.

1. We are showing empty for the participant timestamp.
<img width="1407" alt="Screenshot 2023-10-04 at 11 17 22 PM" src="https://github.com/Expensify/App/assets/3069065/330c4a23-6c79-4f8c-8885-a3f2018b55e6">


2. @puneetlath @lukemorawski I am not sure what will the sender info look like in messages

I updated the accountId by keeping displayName blank, based on the `Onyx.merge` comment, but we do receive the participant info in the chat. Hence we see `Hidden` in the header, but the actual email address in the chat history.

<img width="734" alt="Screenshot 2023-10-04 at 11 20 57 PM" src="https://github.com/Expensify/App/assets/3069065/f9ab6cb3-719a-4bb9-8717-4fe1e20fcb5a">


3. When we're typing from the Hidden account, it shows `Someone is typing`. This definitely looks better than `Hidden is typing`, but wanted to point that's what is expected here.

https://github.com/Expensify/App/assets/3069065/8be2a391-b4d7-4738-87e6-d9fbf7b17a70

4. @puneetlath should we be able to assign tasks to these accounts? We don't show `Hidden` accounts when we search.

5. When transaction is paid we show `Hidden paid ₹24.00 elsewhere`. 

6. When viewing the thread with the Hidden account, the Members show `Hidden` for the title as well as subtitle. Should we remove the subtitle here?

<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/3069065/02640128-2198-421f-86bc-b37261cdef2e">

7. The GH issue mentions to fallback on `Hidden` when neither displayName nor login is available. I am unable to figure out if this would impact the `Split Bill` test.

https://github.com/Expensify/App/blob/b739088fdac88ba4f47ec573229f4d4995c963ef/src/libs/actions/IOU.js#L1040C31-L1040C31

This line uses `participant.login` and if we don't pass this it basically throws an error.

https://github.com/Expensify/App/assets/3069065/001b94cf-fe5f-489e-952b-d0e8a1fd6a04

8. @lukemorawski  when we do a Request money with Scan, the optimistic data is empty and hence the UI shows ` & member-name`. 

<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/3069065/487830af-06e7-4501-a382-85acc0f8edd4">

lukemorawski commented:
@mananjadhav Great work! I want to wait for @puneetlath input on some of your findings and then I will address them.

puneetlath commented:
Yes, great questions! 

>Also do we have a real account that we can replicate for this behavior? Basically updating the Onyx every now and then doesn't seem very reliable.

Unfortunately, no. We always send the displayName from the back-end right now and we are waiting for the updates here before we stop sending it. So we should just catch as much as we can now and then if there are further bugs we can fix them once the back-end stops sending. 

>We are showing empty for the participant timestamp.

Let's not show the timestamp string at all when there is no displayName. 

>@puneetlath @lukemorawski I am not sure what will the sender info look like in messages

Once they send you a message, you should receive their displayName. The only scenario when you won't have it is when you've messaged someone you don't know and they haven't messaged you back. So I don't think we need to worry about this. 

>When we're typing from the Hidden account, it shows Someone is typing. This definitely looks better than Hidden is typing, but wanted to point that's what is expected here.

Someone is typing sounds good, so let's leave that as-is. 

>@puneetlath should we be able to assign tasks to these accounts? We don't show Hidden accounts when we search.

No, I think it's fine as-is in not being able to assign tasks to these accounts. 

>When transaction is paid we show Hidden paid ₹24.00 elsewhere.

Similar to sending a message, you should receive their displayName from the back-end when they pay. (if not we, can fix that from the back-end) 

>When viewing the thread with the Hidden account, the Members show Hidden for the title as well as subtitle. Should we remove the subtitle here?

I think it's fine as-is. 

>The GH issue mentions to fallback on Hidden when neither displayName nor login is available. I am unable to figure out if this would impact the Split Bill test.

Hmm, if you type the whole email address when splitting the bill we should use that. Otherwise, you shouldn't be able to split with them. 

>@lukemorawski when we do a Request money with Scan, the optimistic data is empty and hence the UI shows  & member-name.

Let's fall back on "hidden" here.

mananjadhav commented:
Thanks for the clarification @puneetlath. 

@lukemorawski I think we have all the responses and we can fix the agreed bugs. We also have merge conflicts.

puneetlath commented:
@lukemorawski mind providing an updated ETA?

lukemorawski commented:
@puneetlath Hey! Will be working on that tomorrow.

puneetlath commented:
@lukemorawski friendly bump for an update 😄

lukemorawski committed:
iou display name for report action item when hidden

lukemorawski committed:
fix/dont allow users without login to split payment/money request

lukemorawski committed:
dont show participants time when there's no display name

lukemorawski committed:
Merge branch 'main' into 27393-displayName_fallback_to_hidden

lukemorawski committed:
linting

lukemorawski commented:
@puneetlath 
fixed:
- showing hidden on money request participant item
- not showing the local time when participants displayName is absent
- removed participants without `login` upon split bill - though one comment here. Maybe the app should allow to split bill with a Hidden user, but instead of `login` the backend should accept here the `accountID` ?

mananjadhav commented:
@lukemorawski thanks for the changes. Did 1 pass of the review, will finish this tomorrow. 

> when we do a Request money with Scan, the optimistic data is empty and hence the UI shows & member-name.

Can you confirm if we fixed this one from the previous list.

lukemorawski commented:
@mananjadhav Yep, should work now.

puneetlath commented:
@mananjadhav I just want to make sure this is high on your list. I'm eager to get this wrapped up.

mananjadhav commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<img width="1512" alt="web-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/10b98202-deaa-49d7-badb-9f8ee457b15d">

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<img width="362" alt="mweb-chrome-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/e45bc098-4f3f-49e0-a4e3-c135fca38d3f">

</details>

<details>
<summary>Mobile Web - Safari</summary>

<img width="430" alt="mweb-safari-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/ac90c8c4-c913-45d8-96c9-ca6819f1ab1e">

</details>

<details>
<summary>Desktop</summary>

<img width="1199" alt="desktop-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/f2ff9f72-eb35-4800-b84b-4f0290f87511">

</details>

<details>
<summary>iOS</summary>

<img width="406" alt="ios-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/2c42c54c-b16f-449a-a8de-ae15106c106b">

</details>

<details>
<summary>Android</summary>

<img width="370" alt="android-hidden-display-name" src="https://github.com/Expensify/App/assets/3069065/f3d97378-485e-4d23-9227-3c6f0474b747">

</details>


---

@puneetlath All set. This tests well now. Took a bit longer to retest everything. Also I am unable to overwrite the Onyx data on native apps, hence I've uploaded the basic screenshots there, but overall it will work as I've manually set via code and checked once.

puneetlath commented on src/libs/OptionsListUtils.js:
```diff
@@ -1215,6 +1215,7 @@ function getOptions(
         };
         userToInvite = createOption([optimisticAccountID], personalDetailsExtended, null, reportActions, {
             showChatPreviewLine,
+            shouldNotFallbackToHidden: true,
```
```suggestion
            shouldFallbackToHidden: true,
```

Right? That's what it seems to be in the function above.

lukemorawski replied:
Because it should fall back to hidden, the exception to that rule implemented somewhere else.

puneetlath replied:
I don't see `shouldNotFallbackToHidden` used anywhere else in the codebase. I only see `shouldFallbackToHidden`. So what will passing `shouldNotFallbackToHidden` here do?

puneetlath reviewed (changes requested):
Just two comments!

lukemorawski commented:
@puneetlath well, i don't know exactly why, but when `getDisplayNameForParticipant` is called and it's a new account (meaning, that email address is not present in expensify db) `getPersonalDetailsForAccountID` returns an object that contains only `avatar` key. That how a new/invite account can be detected withing the `getDisplayNameForParticipant` and prevent fallback to "Hidden"

puneetlath commented:
Ahh ok, I see. What if instead of creating this new `hasOnlyAvatarField` function, we were to just update [getPersonalDetailsForAccountID](https://github.com/Expensify/App/blob/e8dcbdba09ae0ffedda838b86da5d76dc6d125ae/src/libs/ReportUtils.js#L1152) like so: 

```diff
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -1164,6 +1164,7 @@ function getPersonalDetailsForAccountID(accountID) {
     return (
         (allPersonalDetails && allPersonalDetails[accountID]) || {
             avatar: UserUtils.getDefaultAvatar(accountID),
+            isOptimisticPersonalDetail: true,
         }
     );
```

And then in `getDisplayNameForParticipant` instead of checking for `hasOnlyAvatarField` we check for `isOptimisticPersonalDetail`.

puneetlath commented:
@mananjadhav @lukemorawski thoughts on my suggestion?

lukemorawski committed:
removed shouldFallbackToHidden flag from createOption method

lukemorawski committed:
refactored check for invite account

puneetlath commented on src/libs/ReportUtils.js:
```diff
@@ -1173,27 +1174,36 @@ function getPersonalDetailsForAccountID(accountID) {
         return '';
     }
     const personalDetails = getPersonalDetailsForAccountID(accountID);
+    // check if it's an invite account
```
Let's make this comment more descriptive to explain why we are doing this, rather than just saying what we're doing.

lukemorawski replied:
@puneetlath comment updated

puneetlath replied:
Great, thanks!

puneetlath reviewed (changes requested):
Looks good to me! 

One small comment and @mananjadhav it'd be great to do a quick re-test.

mananjadhav commented:
I'll retest this by eod.

lukemorawski committed:
commment update

puneetlath commented:
Great, good with me too!

@lukemorawski looks like you have a Jest test failing though. Not sure if that is because of your changes or if you just need to merge main. Once you fix that, we should be good to merge.

lukemorawski committed:
Merge branch 'main' into 27393-displayName_fallback_to_hidden

lukemorawski committed:
getDisplayNameForParticipant optimistic personal details improvements

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

situchan commented:
This PR caused serious regression - [Not able to submit distance request at all](https://github.com/Expensify/App/pull/29843#issuecomment-1771671836).
I confirmed that is not reproducible after reverting this PR locally.
@lukemorawski can you please raise quick PR?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6592629328) to staging by https://github.com/puneetlath in version: 1.3.88-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6608779761) to production by https://github.com/chiragsalian in version: 1.3.88-11 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

s77rt commented on src/pages/iou/steps/MoneyRequestConfirmPage.js:
```diff
@@ -70,10 +70,13 @@ function MoneyRequestConfirmPage(props) {
-                const isPolicyExpenseChat = lodashGet(participant, 'isPolicyExpenseChat', false);
-                return isPolicyExpenseChat ? OptionsListUtils.getPolicyExpenseReportOption(participant) : OptionsListUtils.getParticipantsOption(participant, props.personalDetails);
-            }),
+            _.chain(props.iou.participants)
+                .map((participant) => {
+                    const isPolicyExpenseChat = lodashGet(participant, 'isPolicyExpenseChat', false);
+                    return isPolicyExpenseChat ? OptionsListUtils.getPolicyExpenseReportOption(participant) : OptionsListUtils.getParticipantsOption(participant, props.personalDetails);
+                })
+                .filter((participant) => !!participant.login)
+                .value(),
```
@lukemorawski Can you please clarify why we need to filter participants on login? cc @mananjadhav

s77rt replied:
@lukemorawski Thanks for your reply. It looks like making money requests to "Hidden" accounts works and it's actually a wanted functionality. Since this filter was only done to avoid an error that no longer occurs I think it's safe to remove it. Let us know if disagree.
